### PR TITLE
[firebase_auth] Add macOS support

### PR DESCRIPTION
**Pull request by cbenhagen**
## Description

Add macOS support.

## Related Issues

https://github.com/FirebaseExtended/flutterfire/issues/1653

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] If the pull request affects only one plugin, the PR title starts with the name of the plugin in brackets (e.g. [cloud_firestore])
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/FirebaseExtended/flutterfire/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://www.dartlang.org/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

**Comment by AndroidNils** [comment]
Great work! Is here any process?

**Comment by cbenhagen** [comment]
I shared the code between iOS and macOS by hiding unsupported stuff. Here is a diff for convenience as git will not show it due to the symlink:
```diff
--- FLTFirebaseAuthPlugin-old.m	2020-02-16 12:54:25.000000000 +0100
+++ FLTFirebaseAuthPlugin.m	2020-02-16 12:55:36.000000000 +0100
@@ -1,6 +1,7 @@
 // Copyright 2017 The Chromium Authors. All rights reserved.
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
+#include <TargetConditionals.h>
 
 #import "FLTFirebaseAuthPlugin.h"
 #import "UserAgent.h"
@@ -43,7 +44,9 @@
   FLTFirebaseAuthPlugin *instance = [[FLTFirebaseAuthPlugin alloc] init];
   instance.channel = channel;
   instance.authStateChangeListeners = [[NSMutableDictionary alloc] init];
+#if TARGET_OS_IPHONE
   [registrar addApplicationDelegate:instance];
+#endif
   [registrar addMethodCallDelegate:instance channel:channel];
 
   SEL sel = NSSelectorFromString(@"registerLibrary:withVersion:");
@@ -69,6 +72,7 @@
   return [FIRAuth authWithApp:[FIRApp appNamed:appName]];
 }
 
+#if TARGET_OS_IPHONE
 - (bool)application:(UIApplication *)application
     didReceiveRemoteNotification:(NSDictionary *)notification
           fetchCompletionHandler:(void (^)(UIBackgroundFetchResult result))completionHandler {
@@ -87,6 +91,7 @@
 - (BOOL)application:(UIApplication *)app openURL:(NSURL *)url options:(NSDictionary *)options {
   return [[FIRAuth auth] canHandleURL:url];
 }
+#endif
 
 // TODO(jackson): We should use the renamed versions of the following methods
 // when they are available in the Firebase SDK that this plugin is dependent on.
@@ -265,7 +270,9 @@
                                                          forObject:nil
                                                              error:error];
                                                 }];
-  } else if ([@"updatePhoneNumberCredential" isEqualToString:call.method]) {
+  }
+#if TARGET_OS_IPHONE
+  else if ([@"updatePhoneNumberCredential" isEqualToString:call.method]) {
     FIRPhoneAuthCredential *credential =
         (FIRPhoneAuthCredential *)[self getCredential:call.arguments];
     [[self getAuth:call.arguments].currentUser
@@ -273,7 +280,9 @@
                          completion:^(NSError *_Nullable error) {
                            [self sendResult:result forObject:nil error:error];
                          }];
-  } else if ([@"updatePassword" isEqualToString:call.method]) {
+  }
+#endif
+  else if ([@"updatePassword" isEqualToString:call.method]) {
     NSString *password = call.arguments[@"password"];
     [[self getAuth:call.arguments].currentUser updatePassword:password
                                                    completion:^(NSError *error) {
@@ -332,7 +341,9 @@
                                                    identifier.intValue]
                 details:nil]);
     }
-  } else if ([@"verifyPhoneNumber" isEqualToString:call.method]) {
+  }
+#if TARGET_OS_IPHONE
+  else if ([@"verifyPhoneNumber" isEqualToString:call.method]) {
     NSString *phoneNumber = call.arguments[@"phoneNumber"];
     NSNumber *handle = call.arguments[@"handle"];
     [[FIRPhoneAuthProvider provider]
@@ -367,7 +378,9 @@
                                        forAuthDataResult:authResult
                                                    error:error];
                                  }];
-  } else if ([@"setLanguageCode" isEqualToString:call.method]) {
+  }
+#endif
+  else if ([@"setLanguageCode" isEqualToString:call.method]) {
     NSString *language = call.arguments[@"language"];
     [[self getAuth:call.arguments] setLanguageCode:language];
     [self sendResult:result forObject:nil error:nil];
@@ -487,13 +500,17 @@
   } else if ([FIRGitHubAuthProviderID isEqualToString:provider]) {
     NSString *token = data[@"token"];
     credential = [FIRGitHubAuthProvider credentialWithToken:token];
-  } else if ([FIRPhoneAuthProviderID isEqualToString:provider]) {
+  }
+#if TARGET_OS_IPHONE
+  else if ([FIRPhoneAuthProviderID isEqualToString:provider]) {
     NSString *verificationId = data[@"verificationId"];
     NSString *smsCode = data[@"smsCode"];
     credential = [[FIRPhoneAuthProvider providerWithAuth:[self getAuth:arguments]]
         credentialWithVerificationID:verificationId
                     verificationCode:smsCode];
-  } else if ([provider length] != 0 && data[@"idToken"] != (id)[NSNull null] &&
+  }
+#endif
+  else if ([provider length] != 0 && data[@"idToken"] != (id)[NSNull null] &&
              (data[@"accessToken"] != (id)[NSNull null] ||
               data[@"rawNonce"] != (id)[NSNull null])) {
     NSString *idToken = data[@"idToken"];
```

`addApplicationDelegate` depends on https://github.com/flutter/flutter/issues/41471.

@AndroidNils this is almost finished and if you want to play with it today you can use it with:
```yaml
dependency_overrides:
  firebase_auth:
    git:
      url: https://github.com/cbenhagen/flutterfire
      path: packages/firebase_auth/firebase_auth
      ref: firebase_auth_macos
```

**Comment by AndroidNils** [comment]
> @AndroidNils this is almost finished and if you want to play with it today you can use it with:
> 
> ```yaml
> dependency_overrides:
>   firebase_auth:
>     git:
>       url: https://github.com/cbenhagen/flutterfire
>       path: packages/firebase_auth/firebase_auth
>       ref: firebase_auth_macos
> ```

Thank you! I tested it, but I got this error:
```
Launching lib/main.dart on macOS in debug mode...
** BUILD FAILED **


The following build commands failed:
	CompileC /Users/nils/Desktop/Projekte/myapp/app/build/macos/Build/Intermediates.noindex/Pods.build/Debug/firebase_auth.build/Objects-normal/x86_64/FLTFirebaseAuthPlugin.o /usr/local/Caskroom/flutter/latest/flutter/.pub-cache/git/flutterfire-eff12e2159829287315dcf941512017812767db5/packages/firebase_auth/firebase_auth/darwin/Classes/FLTFirebaseAuthPlugin.m normal x86_64 objective-c com.apple.compilers.llvm.clang.1_0.compiler
(1 failure)
Build process failed
```

My Flutter Version is `v1.15.4-pre.82`

**Comment by cbenhagen** [comment]
Maybe `UserAgent.h` was not generated. Running your app with `flutter run -d macos --verbose`  would show more.

You can try to go into the macos directory and run `pod deintegrate && pod install`.

**Comment by AndroidNils** [comment]
> Maybe `UserAgent.h` was not generated. Running your app with `flutter run -d macos --verbose` would show more.
> 
> You can try to go into the macos directory and run `pod deintegrate && pod install`.

Thanks for the quick response! I tried it, but it didn't work. But I tested it in a new flutter project with the same Firebase Project, etc. and there it worked. So it issue is related to our flutter project. I tried also to move the macOS folder from the working flutter project to our project. But this changed nothing. Some issue.

**Comment by cbenhagen** [comment]
@AndroidNils can you post the output of `flutter run -d macos --verbose`? Maybe add it to a gist and link it here as the output will be quite large.

File: packages/firebase_auth/firebase_auth/darwin/Classes/FLTFirebaseAuthPlugin.m [hunk 1]
```diff
@@ -0,0 +1,543 @@
...
+int nextHandle = 0;
+
++ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar> *)registrar {
+  FlutterMethodChannel *channel =
+      [FlutterMethodChannel methodChannelWithName:@"plugins.flutter.io/firebase_auth"
+                                  binaryMessenger:[registrar messenger]];
+  FLTFirebaseAuthPlugin *instance = [[FLTFirebaseAuthPlugin alloc] init];
+  instance.channel = channel;
+  instance.authStateChangeListeners = [[NSMutableDictionary alloc] init];
+#if TARGET_OS_IPHONE
```
[the remark this answers is not in the document]
Comment (cbenhagen) [reply packages/firebase_auth/firebase_auth/darwin/Classes/FLTFirebaseAuthPlugin.m]: Done

**Comment by AndroidNils** [comment]
> verbose

Here is the log: https://hastebin.com/fopeqovoqi.sql - I hope it helps!

**Comment by cbenhagen** [comment]
@AndroidNils you are indeed missing `UserAgent.h`.  Maybe @collinjackson has an idea generation of this header can be improved.

**Comment by cbenhagen** [comment]
@AndroidNils can you try to:
1. `flutter clean`
2. `cd macos`
3. `rm Podfile.lock`
4.  `pod deintegrate && pod install`

**Comment by cbenhagen** [comment]
@AndroidNils the commands I just sent you most likely will not work. I hope someone has an idea how to properly handle this. Meanwhile you can just copy the `UserAgent.h` from the `macos` directory. In your case this will let you use the plugin:
```
cp /usr/local/Caskroom/flutter/latest/flutter/.pub-cache/git/flutterfire-efe4dca1171ca013e41ec54d3e249cede70a33ae/packages/firebase_auth/firebase_auth/macos/Classes/UserAgent.h /usr/local/Caskroom/flutter/latest/flutter/.pub-cache/git/flutterfire-efe4dca1171ca013e41ec54d3e249cede70a33ae/packages/firebase_auth/firebase_auth/darwin/Classes/UserAgent.h
```

**Comment by AndroidNils** [comment]
> @AndroidNils the commands I just sent you most likely will not work. I hope someone has an idea how to properly handle this. Meanwhile you can just copy the `UserAgent.h` from the `macos` directory. In your case this will let you use the plugin:
> 
> ```
> cp /usr/local/Caskroom/flutter/latest/flutter/.pub-cache/git/flutterfire-efe4dca1171ca013e41ec54d3e249cede70a33ae/packages/firebase_auth/firebase_auth/macos/Classes/UserAgent.h /usr/local/Caskroom/flutter/latest/flutter/.pub-cache/git/flutterfire-efe4dca1171ca013e41ec54d3e249cede70a33ae/packages/firebase_auth/firebase_auth/darwin/Classes/UserAgent.h
> ```

Yes, with this command it works.

**Pull request merged**